### DISCUSSION TITLE
Dirty hack to run Pikaraoke on your Android phone

### DISCUSSION THREAD

**Discussion by jepes1981** [vicwomg/pikaraoke#421] [Show and tell]
### Dirty hack to make pikaraoke work on Android phone using Termux 

(_Termux is an Android terminal emulator and Linux environment app that works directly with no rooting or setup required. A minimal base system is installed automatically - additional packages are available using the APT package manager_.)

download Termux from F-Droid:
https://f-droid.org/en/packages/com.termux/

_*The google play version is no longer the original termux. See_ https://github.com/termux-play-store

open the termux app.

Install python:
`pkg install python`

Follow the standard installation process of PiKaraoke.
I recommend the development way of installing it which is using `poetry` so that you can edit the code more easily.

If you installed it via pip (and you have created an environment) via `python -m venv venv `) you can find the code under:
`venv/lib/python3.12/site-packages/pikaraoke/`

### You will need to modify some codes.

modify lib/get_platform.py

```
def get_platform():
    if sys.platform == "darwin":
        return "osx"
    elif is_raspberry_pi():
        try:
            with open("/proc/device-tree/model", "r") as file:
                model = file.read().strip()
                if "Raspberry Pi" in model:
                    return model  # Returns something like "Raspberry Pi 4 Model B Rev 1.2"
        except FileNotFoundError:
            return "Raspberry Pi - unrecognized"
	    // this is the error encountered when running on termux
		// for some reason, it crashes when this error is not handled
        except PermissionError:    
            print(PermissionError)
            return "unknown"
        
    elif sys.platform.startswith("linux"):
        return "linux"
    elif sys.platform.startswith("win"):
        return "windows"
    else:
        return "unknown"
```

modify karaoke.py  [line 151]

        # Generate connection URL and QR code,
        if self.raspberry_pi:
            # retry in case pi is still starting up
            # and doesn't have an IP yet (occurs when launched from /etc/rc.local)
            end_time = int(time.time()) + 30
            while int(time.time()) < end_time:
				//change the hostname command parameter from uppercase I to lowercase i else it throw an error (see man hostname)
				addresses_str = check_output(["hostname", "-i"]).strip().decode("utf-8", "ignore")
				#addresses_str = check_output(["hostname", "-I"]).strip().decode("utf-8", "ignore")
                addresses = addresses_str.split(" ")
                self.ip = addresses[0]
                if not self.is_network_connected():
                    logging.debug("Couldn't get IP, retrying....")
                else:
                    break
        else:
            self.ip = self.get_ip()

Modify `app.py [around line 553] `so that it will not crash when we go to the settings/try to login.

```
@app.route("/info") 
def info():
    url = k.url

    # cpu
    try:
        cpu = str(psutil.cpu_percent()) + "%"
    except:
        cpu = "unsupported"
```

We now need to install some packages in termux. This is because pikaraoke expects `iwconfig `to be available since `get_platform `assumed unknown/raspberry_pi.

istall first:
`pkg install root-repo
`
then:
`pkg install wireless-tools
`
_*might fail if done in the same command_

if you forgot to install ffmpeg:
`pkg install ffmpeg
`

The hostname command in karaoke.py only returns the loopback address regardless if you are connected to the internet.
We will need to supply the actual ip of your phone else video will not play.

Use the command `ifconfig `in termux, your ip will be under wlan0, inet

As previously mentioned, we need to supply the actual ip, IT IS NECESSARY to use the `--url` argumant/parameter when launching indicating the ip and port of your phone
the `--headless` argument/parameter is also NECESSARY since termux will not launch the chome app.
sample command:
`pikaraoke --url 192.168.0.123:5555 --headless`

Open the headless url on a chrome based browser (on the same phone, another phone, or computer, or smart tv)

### my setup:
phone connected to a bluetook speaker with microphone. I can now sing with all my heart without booting up my computer.
Also, since I always carry my phone, and bluetooth speakers with mic are popular in our area, karaoke is almost instant.

**Comment by vicwomg** [comment]
Fascinating. So that means ffmpeg is installed by default on Android? How did you install pikaraoke? Via pip?

The proper way to do this is modify `get_platform` to properly detect android and not recognize it as a raspberry pi. I imagine since android has ARM processor, it's getting misconfigured. I would welcome a pull request that fixes that.

**Reply by jepes1981** [reply]
> Fascinating. So that means ffmpeg is installed by default on Android? How did you install pikaraoke? Via pip?
> 

I actually forgot about ffmpeg since I explicitly installed it previously on my termux app. I have updated the steps on my post in installing ffmpeg. I have also included the code modification so that it won't crash when it tries to check the cpu usage.

I installed it via pip. Edited the code via sshd (I will never edit code on phone, too small, touch keyboard sucks lol). Termux is the closest to a linux distro as an android app 

> The proper way to do this is modify `get_platform` to properly detect android and not recognize it as a raspberry pi. I imagine since android has ARM processor, it's getting misconfigured. I would welcome a pull request that fixes that.

I agree but it will be finicky. Android is very restrictive of apps accessing what an app should only access (tasker almost got delisted because of this). Same reason why termux dropped their google play uploads and sticked with f-droid. The method failed because it tried to read` /proc/device-tree/model` where normally you can view and read the contents, in android, you are not even allowed to view/list the root directory, only the path where the app was installed and has to explicitly allow access to user storage. The way I did it was only to bypass the error but it can actually be checked by explicitly testing if it is running on termux by checking the `$PREFIX` env variable if it contains the string `com.termux`.

I am planning to make a pull request to fix that (and support termux) but I am not making any promises anytime soon since this is just a band-aid to make it work, not proper code (The primary reason I posted this so that I can still remember what I did). I will need to do the following:

- proper way to detect if running in termux (fix available)
- properly getting ip address instead of needing to manually get the ip from terminal/device and explicitly passing it via `--url `
- autmatically default to headless when detected running on termux app.
- proper logic/alternative to iwconfig when running on termux to avoid having to install root-repo and wireless-tools (since they don't acutally work when your phone is not rooted)
- CPU stats will remain unsupported since there in no way to check cpu usage on android if you are not rooted.

**Comment by jepes1981** [comment]
created PR: https://github.com/vicwomg/pikaraoke/pull/423

**Comment by vicwomg** [comment]
Merged! Will release it today and have added your guide to wiki

https://github.com/vicwomg/pikaraoke/wiki/Running-on-Android-using-termux

**Comment by vicwomg** [comment]
@jepes1981 I just tried this myself on my 8-year old Oneplus 3 . Works great! I'm shocked.

A couple of things I noticed though:
- Transposing / changing song key doesn't work. My guess is the lib-rubberband isn't bundled with this version of ffmpeg. Might need to manually compile it and add (found this: https://stackoverflow.com/questions/44795258/ffmpeg-on-android-with-rubberband )
- If the phone screen goes off the app stops responding to commands. My guess is android suspends the network when the screen is off.
- Installing all the apt updates took a really long time. It was downloading from a mirror in Asia. I realized I could switch my repository using `termux-change-repo`. I added this to the wiki

**Reply by jepes1981** [reply]
Re transposing. most probably that is the case. Either indicate that transposing is umavailable in Termux or ask support to the ffmpeg maintainer especially why rubberband was dropped where it can actually be compiled with. I'll need to add some debug logging to catch the issue when I get back. I'll also try to tackle the issue on videos not playing when not connected to a network.


If the phone screen goes off the app stops responding to commands. My guess is android suspends the network when the screen is off.
Re app stops responding to commands when screen goes off. Its not just screen going off, it may also happen when placing termux in the background. It is an inherent issue with android espcially if you are on android v12. This can be mitigated, with less success tbh, by enabling acquire wake lock, or in my case, my phone has feature to lock app in memory. Similar solution can be found in https://tasker.joaoapps.com/userguide/ru/androidpowermanagement.html


>Installing all the apt updates took a really long time. It was downloading from a mirror in Asia. I realized I could switch my repository using termux-change-repo. I added this to the wiki

Yes, this can help a lot.

**Reply by jepes1981** [reply]
> @jepes1981 I just tried this myself on my 8-year old Oneplus 3 . Works great! I'm shocked.
> 
> A couple of things I noticed though:
> 
> * Transposing / changing song key doesn't work. My guess is the lib-rubberband isn't bundled with this version of ffmpeg. Might need to manually compile it and add (found this: https://stackoverflow.com/questions/44795258/ffmpeg-on-android-with-rubberband )

Confirmed that the issue was lib-rubberband isn't bundled with the version ffmpeg. (refer to https://github.com/termux/termux-packages/blob/master/packages/ffmpeg/build.sh and https://ffmpeg.org/ffmpeg-filters.html#rubberband).
I was able to successfully recompile ffmpeg with  --enable-librubberband by editing the build.sh using the on-device usage (https://wiki.termux.com/wiki/Building_packages) in just 3 commands but the build time takes about several minutes. This is not recommended for the average user.  
I'll explore if ffmpeg can be patched post install since they have not updated ffmpeg for a while (v6 for termux, v7  is the latest).

I tried bypassing tramsposition by adding the forcing is_transposed back to false in karaoke.py if it is on Android:
```
        # copy the audio stream if no transposition/normalization, otherwise reincode with the aac codec
        is_transposed = semitones != 0
        if self.platform == "android":
            logging.warning("lib-rubberband is not supported by the default ffmpeg install in Termux")
            is_transposed = False
```
This just ignore the semitone changes and prevents pikaraoke from hanging/freezing/looping.
I would have tried changing it in the user interface side right now  but tbh, i'm no expert in python and i'm a total noob in flask/web-dev. It will take a while for me to modify the flask/web-ui part.

**Reply by vicwomg** [reply]
is it possible to submit a PR to the termux-packages to add the library to ffmpeg? Or write a guide on how to compile ffmpeg to add to the wiki? If a user is running termux on android, I would assume they are more advanced than the average user.

Ideally we wouldn't need to hard-code disabling features that do not need to be disabled.

**Reply by jepes1981** [reply]
Sure thing. I think I can do both.

**Comment by jepes1981** [comment]
If I mentioned 3 commands, my apologies since I forgot (again) that I have some packages already installed. Below is the steps to re-compile ffmpeg inside Termux, feel free to make any changes as you see fit:


### Key Change / Pitch Shifting not working solution
lib-rubberband is not compiled with ffmpeg resulting to the issue. The solution is to re-compile ffmpeg with --enable-librubberband parameter. It can be done inside Termux with a few commands and a little bit of patience.

**Optional but highly recommended if you will use your on-screen keyboard to edit files is to use "Unexpected Keyboard" as your keyboard app (https://play.google.com/store/apps/details?id=juloo.keyboard2&hl=en)*

Download git if it is not yet installed:
```
pkg install git
```
You will also need a text editor (e.g. vim or nano, or what ever your choice is):

```
pkg install nano
```
or
```
pkg install vim
```
**please don't get stuck in vim limbo*
https://www.reddit.com/r/linux4noobs/comments/p5i418/how_do_i_quit_vim/ 

Install rubberband package:
```
pkg install rubberband
```

Clone the termux-packages repository:
```
git clone https://github.com/termux/termux-packages.git
```
cd into `termux-packages`
```
cd termux-packages
```
Prepare Termux for building packages (this can take a few minutes to run):
```
./scripts/setup-termux.sh
```
edit the file `termux-packages/ffmpeg/build.sh`

If you have ffmpeg already installed, you may uninstall it by:
```
pkg uninstall ffmpeg
```

In the file build.sh, under `$TERMUX_PKG_SRCDIR/configure` , add `--enable-librubberband`
Re-compile ffmpeg (compilation will take several minutes*):
````
./build-package.sh -I ffmpeg
````
If there were no errors, ffmpeg should be installed with lib-rubberband support and Key Change / Pitch Shifting should now work.
If the re-compile fails, you may try running the following again:
```
./scripts/setup-termux.sh
./build-package.sh -I ffmpeg
```
*References:
https://wiki.termux.com/wiki/Building_packages
https://github.com/termux/termux-packages/wiki/Building-packages#building-packages*

**Reply by jepes1981** [reply]
Termux packages PR: https://github.com/termux/termux-packages/pull/22384
I just hope it passes their strict (not really, but i'm new) guideline. 😅

**Reply by jepes1981** [reply]
I will be retracting the PR in Termux since lib-rubberband was intentionally dropped from ffmpeg due to licensing. That means if FFmpeg did not included in their binaries, Termux cannot include it in their distribution as well (to my understanding).

packages/ffmpeg/build.sh
```
# GPLed FFmpeg binaries linked against fdk-aac are not redistributable.
```

Also refer to https://github.com/termux/termux-packages/?tab=License-1-ov-file

**Reply by vicwomg** [reply]
Your build instructions worked great, I have updated the wiki with them. It did take a very long time, but I have transposition working now. 

ffmpeg not coming with librubberband seems like it could affect more than just termux. Pikaraoke should probably be able to detect if librubberband is missing and disable transposition if so, as well as ffmpeg and chrome in general. I will write up a task for that.

**Reply by jepes1981** [reply]
Hold up on your code update. I just received confirmation to proceed with linking lib-rubberband as long as I don't enable libfdk-aac in the build.  Thou It can still be helpful on supporting unknown distribution where librubberband is not linked to ffmpeg.

**Reply by vicwomg** [reply]
My thinking is handling the latter case. I've encountered other distributions where librubberband was not included and it would be great to handle them. It's not super high priority. I don't think I'll get to it soon.

**Reply by jepes1981** [reply]
> My thinking is handling the latter case. I've encountered other distributions where librubberband was not included and it would be great to handle them. It's not super high priority. I don't think I'll get to it soon.

I think I have a few ideas how. It's an opportunity for to familarize myself making forks and PRs.

**Comment by jepes1981** [comment]
My PR was approved and merged regarding enabling ligrubberbamd on ffmpeg. 

https://github.com/termux/termux-packages/pull/22384

**Reply by jepes1981** [reply]
```
ffmpeg -filters | grep rubberband
ffmpeg version 6.1.2 Copyright (c) 2000-2024 the FFmpeg developers
  built with Android (12470979, +pgo, +bolt, +lto, +mlgo, based on r522817c) clang version 18.0.3 (https://android.googlesource.com/toolchain/llvm-project d8003a456d14a3deb8054cdaa529ffbf02d9b262)
  configuration: --arch=aarch64 --as=aarch64-linux-android-clang --cc=aarch64-linux-android-clang --cxx=aarch64-linux-android-clang++ --nm=llvm-nm --ar=llvm-ar --ranlib=llvm-ranlib --pkg-config=/home/builder/.termux-build/_cache/android-r27c-api-24-v1/bin/pkg-config --strip=llvm-strip --cross-prefix=aarch64-linux-android- --disable-indevs --disable-outdevs --enable-indev=lavfi --disable-static --disable-symver --enable-cross-compile --enable-gnutls --enable-gpl --enable-version3 --enable-jni --enable-lcms2 --enable-libaom --enable-libass --enable-libbluray --enable-libdav1d --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libharfbuzz --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenmpt --enable-libopus --enable-librav1e --enable-librubberband --enable-libsoxr --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-mediacodec --enable-opencl --enable-shared --prefix=/data/data/com.termux/files/usr --target-os=android --extra-libs=-landroid-glob --disable-vulkan --enable-neon --disable-libfdk-aac
  libavutil      58. 29.100 / 58. 29.100
  libavcodec     60. 31.102 / 60. 31.102
  libavformat    60. 16.100 / 60. 16.100
  libavdevice    60.  3.100 / 60.  3.100
  libavfilter     9. 12.100 /  9. 12.100
  libswscale      7.  5.100 /  7.  5.100
  libswresample   4. 12.100 /  4. 12.100
  libpostproc    57.  3.100 / 57.  3.100
 ..C rubberband        A->A       Apply time-stretching and pitch-shifting.
```
pikaraoke semitones are now officially supported on Termux.

**Reply by vicwomg** [reply]
Most excellent. Looks like this was merged 2 days ago. Can you confirm the official apt install includes it? If so, I will update the wiki

**Reply by jepes1981** [reply]
> Most excellent. Looks like this was merged 2 days ago. Can you confirm the official apt install includes it? If so, I will update the wiki

it is. see screenshot below:
![Screenshot_2024-12-03-05-56-16-837_com.termux.jpg](https://github.com/user-attachments/assets/b1b35add-21c1-4e94-be52-5cfdac3c2cd4)
